

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Getting started &mdash; Grape 2.0 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="Grape 2.0 documentation" href="index.html"/>
        <link rel="prev" title="Grape 2.0" href="index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> Grape</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="">Getting started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#creating-a-working-directory">Creating a working directory</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reference-files">Reference files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#installing-nextflow">Installing Nextflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pipeline-usage">Pipeline usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#input-format">Input format</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-the-pipeline">Run the pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#stop-the-pipeline">Stop the pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#job-monitoring">Job monitoring</a></li>
</ul>
</li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Grape</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Getting started</li>
      <li class="wy-breadcrumbs-aside">
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="getting-started">
<h1>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h1>
<div class="section" id="creating-a-working-directory">
<h2>Creating a working directory<a class="headerlink" href="#creating-a-working-directory" title="Permalink to this headline">¶</a></h2>
<p>Move to the folder where you want to create the base directory for the pipeline and run the following command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>git clone -b nextflow /users/rg/projects/git/bp.pipeline.git &lt;pipeline name&gt;
<span class="nv">$ </span><span class="nb">cd</span> &lt;pipeline name&gt;
</pre></div>
</div>
</div>
<div class="section" id="reference-files">
<h2>Reference files<a class="headerlink" href="#reference-files" title="Permalink to this headline">¶</a></h2>
<p>Only genome FASTA file and annotation file are required. Reference files can be specified using absolute paths or relative paths to a local folder where links to the files can be created:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>mkdir refs
<span class="nv">$ </span>ln -s &lt;path to the genome file&gt; refs
<span class="nv">$ </span>ln -s &lt;path to the annoation file&gt; refs
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The genome and the annotation needs to be sorted by <strong>genomic position</strong>.</p>
</div>
</div>
<div class="section" id="installing-nextflow">
<h2>Installing Nextflow<a class="headerlink" href="#installing-nextflow" title="Permalink to this headline">¶</a></h2>
<p>This step will install nextflow for your user. It is only needed once. These are the commands:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>curl -fsSL get.nextflow.io | bash
<span class="nv">$ </span>mv nextflow &lt;folder in your path&gt;
<span class="nv">$ </span>nextflow info
Version: 0.12.0 build 2580
Modified: 05-01-2015 15:07 UTC <span class="o">(</span>16:07 CEST<span class="o">)</span>
System: Linux 2.6.32-504.1.3.el6.x86_64
Runtime: Groovy 2.3.9 on Java HotSpot<span class="o">(</span>TM<span class="o">)</span> 64-Bit Server VM 1.7.0_60-b19
Encoding: UTF-8 <span class="o">(</span>UTF-8<span class="o">)</span>
</pre></div>
</div>
</div>
<div class="section" id="pipeline-usage">
<h2>Pipeline usage<a class="headerlink" href="#pipeline-usage" title="Permalink to this headline">¶</a></h2>
<p>To get the usage string and the list of options use this command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>nextflow run blueprint.pipeline.nf --help
N E X T F L O W  ~  version 0.12.0

B L U E P R I N T ~ RNA Pipeline
--------------------------------
Run the RNAseq pipeline on a <span class="nb">set </span>of data.

Usage:
    blueprint.pipeline.nf -i INDEX_FILE -g GENOME_FILE -a ANNOTATION_FILE <span class="o">[</span>OPTION<span class="o">]</span>...

Options:
    --help                              Show this message and exit.
    --index INDEX_FILE                  Index file.
    --genome GENOME_FILE                Reference genome file<span class="o">(</span>s<span class="o">)</span>.
    --annotation ANNOTAION_FILE         Reference gene annotation file<span class="o">(</span>s<span class="o">)</span>.
    --chunk-size                        The number of records to be put in each chunk when splitting the input. Default: no split
    --error-strategy                    Specify how an error condition is managed by the pipeline processes. Default: the entire pipeline
                                        terminates <span class="k">if </span>a process returns an error status.
    --max-read-length READ_LENGTH       The maximum <span class="nb">read </span>length <span class="o">(</span>used to compute the transcriptomes<span class="o">)</span>. Default: <span class="s2">&quot;auto&quot;</span>.
    --max-mismatches THRESHOLD          Set maps with more than &lt;threshold&gt; error events to unmapped. Default <span class="s2">&quot;4&quot;</span>.
    --max-multimaps THRESHOLD           Set multi-maps with more than &lt;threshold&gt; mappings to unmapped. Default <span class="s2">&quot;10&quot;</span>.
    --filter-intron-length THRESHOLD    Filter multimaps preferring ones with intron length &gt; THRESHOLD
    --filter-block-length THRESHOLD     Filter multimaps preferring ones with block length &gt; THRESHOLD
    --filter-level LEVEL                Reduce multimaps using the specified uniqueness level.
    --rg-platform PLATFORM              Platform/technology used to produce the reads <span class="k">for </span>the BAM @RG tag.
    --rg-library LIBRARY                Sequencing library name <span class="k">for </span>the BAM @RG tag.
    --rg-center-name CENTER_NAME        Name of sequencing center that produced the reads <span class="k">for </span>the BAM @RG tag.
    --rg-desc DESCRIPTION               Description <span class="k">for </span>the BAM @RG tag.
    --flux-profile                      Specify whether the Flux Capacitor profile file shoudl be written. Default: <span class="s2">&quot;false&quot;</span>.
    --count-elements ELEMENTS           A comma separated list of elements to be counted by the Flux Capacitor.
                                        Possible values: INTRONS,SPLICE_JUNCTIONS. Defalut: <span class="s2">&quot;none&quot;</span>.
</pre></div>
</div>
</div>
<div class="section" id="input-format">
<h2>Input format<a class="headerlink" href="#input-format" title="Permalink to this headline">¶</a></h2>
<p>The pipeline needs a tab separated file as one of the inputs. This files should contain information about the FASTQ files to be processed. The columns needed in order are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="91%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">sample</span></tt></td>
<td>the sample identifier, used to merge bam files in case multiple runs for the same sample are present</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">id</span></tt></td>
<td>the run identifier (e.g. labExpId)</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">path</span></tt></td>
<td>the path to the fastq file</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">type</span></tt></td>
<td>the type (e.g. fastq)</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">view</span></tt></td>
<td>an attribute that specifies the content of the file (e.g. FastqRd1)</td>
</tr>
</tbody>
</table>
<p>Here is an example:</p>
<div class="highlight-python"><div class="highlight"><pre>sample1  test1   data/test1_1.fastq.gz   fastq   FastqRd1
sample1  test1   data/test1_2.fastq.gz   fastq   FastqRd2
</pre></div>
</div>
<p>Sample and id can be the same in case you don&#8217;t have/know sample identifiers:</p>
<div class="highlight-python"><div class="highlight"><pre>test  test   data/test.fastq.gz   fastq   FastqRd
</pre></div>
</div>
</div>
<div class="section" id="run-the-pipeline">
<h2>Run the pipeline<a class="headerlink" href="#run-the-pipeline" title="Permalink to this headline">¶</a></h2>
<p>Here is a simple example of the command to run the pipeline:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>nextflow -bg run blueprint.pipeline.nf --index input-files.tsv --genome refs/hg38.AXYM.fa --annotation refs/gencode.v21.annotation.AXYM.gtf --rg-platform ILLUMINA --rg-center-name CRG -resume 2&gt;&amp;1 &gt; pipeline.log
</pre></div>
</div>
</div>
<div class="section" id="stop-the-pipeline">
<h2>Stop the pipeline<a class="headerlink" href="#stop-the-pipeline" title="Permalink to this headline">¶</a></h2>
<p>To stop a running pipeline just run the following command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span><span class="nb">kill</span> <span class="k">$(</span>cat .nextflow.pid<span class="k">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If you run multiple pipelines within the same folder (e.g. for different genders), the file <tt class="docutils literal"><span class="pre">.nextflow.pid</span></tt> will contain only the pid of the last Nextflow pipeline (<em>while all pipelines will still be running</em>). In that case you will need to manually look for the pid and kill the corresponding process. Since this approach can be quite confusing it is strongly discouraged for the moment. The problem has been reported and it is likely to be fixed in a next release.</p>
</div>
</div>
<div class="section" id="job-monitoring">
<h2>Job monitoring<a class="headerlink" href="#job-monitoring" title="Permalink to this headline">¶</a></h2>
<p>Nextflow runs all processes in an isolated directory under the pipeline working folder (by default <tt class="docutils literal"><span class="pre">./work</span></tt>). Each process is configured and run by means of several files contained in the process folder. Among those files some can be worth noting:</p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="84%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">.command.env</span></tt></td>
<td>the process environment</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">.command.out</span></tt></td>
<td>the process output (merge <tt class="docutils literal"><span class="pre">stderr</span></tt> into <tt class="docutils literal"><span class="pre">stdout</span></tt> if not redirected)</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">.command.run</span></tt></td>
<td>the script submitted to the cluster (also contains the header with cluster directives)</td>
</tr>
<tr class="row-even"><td><tt class="docutils literal"><span class="pre">.command.sh</span></tt></td>
<td>the actual command</td>
</tr>
<tr class="row-odd"><td><tt class="docutils literal"><span class="pre">.exitcode</span></tt></td>
<td>the exit code of the command</td>
</tr>
</tbody>
</table>
<p>A processe can then be easily monitored by inspecting the process folder. Each process is uniquely represented by a hash string nternally computed by Nextflow using comands and inputs. To inspect a process folder just look for Nextflow submission messages in the pipeline log file, which look like the following:</p>
<div class="highlight-python"><div class="highlight"><pre>...
[b5/0e02e9] Submitted process &gt; index (1)
...
</pre></div>
</div>
<p>The string bewtween square brackets represents the prefix of the relative path to the process folder starting from the <tt class="docutils literal"><span class="pre">work</span></tt> directory. So to inspect the process workiong folder for the <cite>index (1)</cite> process above:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>find work/b5 -name <span class="s1">&#39;0e02e9*&#39;</span> -exec ls -a <span class="o">{}</span> <span class="se">\+</span>
.  ..  .command.begin  .command.env  .command.out  .command.run  .command.sh  .command.val  .exitcode  genome_index.gem  genome_index.log  hg38_AXM.fa
</pre></div>
</div>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="index.html" class="btn btn-neutral" title="Grape 2.0"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, Emilio Palumbo.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'2.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>